File: src/Scripting/Lua/proxy.lua
local RemoteDocumentStore = {}
RemoteDocumentStore.__index = RemoteDocumentStore

function RemoteDocumentStore.new()
  return setmetatable({ fetch_count = 0 }, RemoteDocumentStore)
end

function RemoteDocumentStore:get(id)
  self.fetch_count = self.fetch_count + 1
  return string.format("doc(%d)", id)
end

local DocumentStoreProxy = {}
DocumentStoreProxy.__index = DocumentStoreProxy

function DocumentStoreProxy.new()
  return setmetatable({ backend = nil, cache = {} }, DocumentStoreProxy)
end

function DocumentStoreProxy:get(id)
  local cached = self.cache[id]
  if cached ~= nil then
    return cached
  end

  if self.backend == nil then
    self.backend = RemoteDocumentStore.new()
  end

  local value = self.backend:get(id)
  self.cache[id] = value
  return value
end

function DocumentStoreProxy:backend_count()
  return self.backend == nil and 0 or 1
end

function DocumentStoreProxy:fetch_count()
  return self.backend == nil and 0 or self.backend.fetch_count
end

local store = DocumentStoreProxy.new()
local first = store:get(42)
local second = store:get(42)
print(string.format("backend=%d;fetches=%d;first=%s;second=%s", store:backend_count(), store:fetch_count(), first, second))

File: src/Scripting/PHP/proxy.php
<?php

declare(strict_types=1);

interface DocumentStore
{
    public function get(int $id): string;
}

final class RemoteDocumentStore implements DocumentStore
{
    public int $fetchCount = 0;

    public function get(int $id): string
    {
        $this->fetchCount++;
        return "doc($id)";
    }
}

final class DocumentStoreProxy implements DocumentStore
{
    private ?RemoteDocumentStore $backend = null;
    /** @var array<int, string> */
    private array $cache = [];

    public function get(int $id): string
    {
        if (array_key_exists($id, $this->cache)) {
            return $this->cache[$id];
        }
        $this->backend ??= new RemoteDocumentStore();
        $value = $this->backend->get($id);
        $this->cache[$id] = $value;
        return $value;
    }

    public function backendCount(): int
    {
        return $this->backend === null ? 0 : 1;
    }

    public function fetchCount(): int
    {
        return $this->backend?->fetchCount ?? 0;
    }
}

$store = new DocumentStoreProxy();
$first = $store->get(42);
$second = $store->get(42);
printf("backend=%d;fetches=%d;first=%s;second=%s\n", $store->backendCount(), $store->fetchCount(), $first, $second);
